### ISSUE TITLE
Debian 10 and newest ltfs code with Tandberg lto-6 tape drive does Seqmentation fault

### ISSUE THREAD

**Issue by ia2115** [LinearTapeFileSystem/ltfs#283]
it works with a lto-5 IBM Tape drive, i did the "mod" on supporting TANDBERG DATA into supported drives, on HPE and recompiled..

but gets those Segmentation faults when i run mkltfs and ltfs commands with this tandberg tapedrive attached.

```
root@arc01:~# ltfs -o devname=/dev/sg4 /mnt/ltfs
4f07 LTFS14000I LTFS starting, LTFS version 2.5.0.0 (Prelim), log level 2.
4f07 LTFS14058I LTFS Format Specification version 2.4.0.
4f07 LTFS14104I Launched by "ltfs -o devname=/dev/sg4 /mnt/ltfs".
4f07 LTFS14105I This binary is built for Linux (x86_64).
4f07 LTFS14106I GCC version is 8.3.0.
4f07 LTFS17087I Kernel version: Linux version 4.19.0-16-amd64 (debian-kernel@lists.debian.org) (gcc version 8.3.0 (Debian 8.3.0-6)) #1 SMP Debian 4.19.181-1 (2021-03-19) i386.
4f07 LTFS17089I Distribution: PRETTY_NAME="Debian GNU/Linux 10 (buster)".
4f07 LTFS14063I Sync type is "time", Sync time is 300 sec.
4f07 LTFS17085I Plugin: Loading "sg" tape backend.
4f07 LTFS17085I Plugin: Loading "unified" iosched backend.
4f07 LTFS14095I Set the tape device write-anywhere mode to avoid cartridge ejection.
4f07 LTFS30209I Opening a device through sg-ibmtape driver (/dev/sg4).
Segmentation fault
root@arc01:~# lsscsi -g
[0:0:0:0]    tape    TANDBERG LTO-6 HH         3579  /dev/st0   /dev/sg4
[1:0:0:0]    disk    ATA      Patriot Burst    61.3  /dev/sda   /dev/sg0
[2:0:0:0]    cd/dvd  HL-DT-ST BDDVDRW GGC-H20L 1.02  /dev/sr0   /dev/sg3
[7:0:0:0]    disk    HP       MB2000FAMYV      HPD7  /dev/sdb   /dev/sg1
[7:0:1:0]    disk    HP       MB2000FAMYV      HPD7  /dev/sdc   /dev/sg2
root@arc01:~#
```

```
root@arc01:~/ltfs/src/tape_drivers# ls -la
total 220
drwxr-xr-x 7 root root  4096 May 18 18:52 .
drwxr-xr-x 9 root root  4096 May 18 18:54 ..
-rw-r--r-- 1 root root 10403 May 16 22:26 crc32c_crc.c
-rw-r--r-- 1 root root  2571 May 16 22:26 crc32c_crc.h
drwxr-xr-x 3 root root  4096 May 16 22:26 freebsd
drwxr-xr-x 4 root root  4096 May 16 22:26 generic
-rw-r--r-- 1 root root 15306 May 18 18:51 hp_tape.c
-rw-r--r-- 1 root root  2688 May 16 22:26 hp_tape.h
-rw-r--r-- 1 root root 51073 May 16 22:26 ibm_tape.c
-rw-r--r-- 1 root root  9583 May 16 22:26 ibm_tape.h
drwxr-xr-x 4 root root  4096 May 16 22:26 linux
drwxr-xr-x 3 root root  4096 May 16 22:26 netbsd
-rw-r--r-- 1 root root  6443 May 16 22:26 open_factor.c
-rw-r--r-- 1 root root  2460 May 16 22:26 open_factor.h
drwxr-xr-x 3 root root  4096 May 16 22:26 osx
-rw-r--r-- 1 root root 12579 May 16 22:26 quantum_tape.c
-rw-r--r-- 1 root root  2708 May 16 22:26 quantum_tape.h
-rw-r--r-- 1 root root  7352 May 16 22:26 reed_solomon_crc.c
-rw-r--r-- 1 root root  2498 May 16 22:26 reed_solomon_crc.h
-rw-r--r-- 1 root root  3378 May 16 22:26 spc_op_codes.h
-rw-r--r-- 1 root root  3281 May 16 22:26 ssc_op_codes.h
-rw-r--r-- 1 root root  9029 May 16 22:26 tape_drivers.h
-rw-r--r-- 1 root root 23851 May 16 22:26 vendor_compat.c
-rw-r--r-- 1 root root  3287 May 16 22:26 vendor_compat.h
```

```C
*/
#ifndef mingw_PLATFORM
#if defined (__FreeBSD__) || defined(__NetBSD__)
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#endif /* __FreeBSD__ */
#include <netdb.h>
#include <ifaddrs.h>
#include <unistd.h>

#define LOOP_BACK_DEVICE "lo"
#endif

#include "tape_drivers/hp_tape.h"
#include "libltfs/ltfs_endian.h"

struct supported_device *hp_supported_drives[] = {
                TAPEDRIVE( HP_VENDOR_ID,  "Ultrium 5-SCSI",  DRIVE_LTO5,    "[Ultrium 5-SCSI]" ),  /* HP Ultrium Gen 5  */
                TAPEDRIVE( HP_VENDOR_ID,  "Ultrium 6-SCSI",  DRIVE_LTO6,    "[Ultrium 6-SCSI]" ),  /* HP Ultrium Gen 6  */
                TAPEDRIVE( HP_VENDOR_ID,  "LTO-6 HH        ",  DRIVE_LTO6,    "[TANDBERG DATA LTO6]" ),  /* HP Ultrium Gen 6  */
                TAPEDRIVE( HP_VENDOR_ID,  "Ultrium 7-SCSI",  DRIVE_LTO7,    "[Ultrium 7-SCSI]" ),  /* HP Ultrium Gen 7  */
                TAPEDRIVE( HPE_VENDOR_ID, "Ultrium 8-SCSI",  DRIVE_LTO8,    "[Ultrium 8-SCSI]" ),  /* HPE Ultrium Gen 8 */
                /* End of supported_devices */
                NULL
};

/* HP/HPE LTO tape drive vendor unique sense table */
struct error_table hp_tape_errors[] = {
```

any suggestions?

**Comment by xosevp** [comment]
Try this **UNTESTED** patch: https://paste.debian.net/plain/1198106

do: `$ patch -p1 < 1198106`
recompile it
and run:`# ltfs -o device_list`

**Comment by ia2115** [comment]
i cleaned out my ltfs folder, and git cloned it again.
created this patch file and did what you asked to do.

Recompiled it and installed into debian.
works like a charm, thanks so much xeosevp, really appriciate it. 

i mounted and restored some data, will try format a tape and put data on another one in the weekend.

**Comment by xosevp** [comment]
Please, paste the output `# ltfs -o device_list` here.
Thanks.

**Comment by xosevp** [comment]
It looks good, BUT:

Keep in mind that you are running experimental code from the `master` branch.
The stable repo is at the `v2.4-stable` branch.
And you are the first one to try **Tandberg tape drives**.

You should test that setup thoroughly for a while, before being considered for a production environment.

**Comment by ia2115** [comment]
ye i will thanks a lot for helping out here, and doing this. really appriciated.

**Comment by ia2115** [comment]
Test on the stable-branch commands done as in cmd from post:

# ltfs -o devname=/dev/sa0 /mnt/lto5
LTFS9015W Setting the locale to 'en_US.UTF-8'. If this is wrong, please set the LANG environment variable before starting ltfs.
18d6e LTFS14000I LTFS starting, LTFS version 2.4.3.2 (Prelim), log level 2.
18d6e LTFS14058I LTFS Format Specification version 2.4.0.
18d6e LTFS14104I Launched by "ltfs -o devname=/dev/sa0 /mnt/lto5".
18d6e LTFS14105I This binary is built for FreeBSD.
18d6e LTFS14106I GCC version is FreeBSD Clang 10.0.1 (git@github.com:llvm/llvm-project.git llvmorg-10.0.1-0-gef32c611aa2).
18d6e LTFS17087I Kernel version: FreeBSD 12.2-STABLE r369131 GENERIC
.
18d6e LTFS14063I Sync type is "time", Sync time is 300 sec.
18d6e LTFS17085I Plugin: Loading "cam" tape backend.
18d6e LTFS17085I Plugin: Loading "unified" iosched backend.
18d6e LTFS14095I Set the tape device write-anywhere mode to avoid cartridge ejection.
18d6e LTFS31223I Opening a device through CAM driver (/dev/sa0)
18d6e LTFS31228I Product ID is 'LTO-6 HH'
18d6e LTFS31229I Vendor ID is TANDBERG
18d6e LTFS31232I Firmware revision is 3579
18d6e LTFS31233I Drive serial is HUJ527449P
18d6e LTFS17160I Maximum device block size is 131072.
18d6e LTFS11330I Loading cartridge.
18d6e LTFS31258I Cannot get remaining capacity: failed to parse the log page
18d6e LTFS12030E Cannot get capacity data: backend call failed (-21704).
18d6e LTFS11999E Cannot load the medium: failed to get capacity data (-21704).
18d6e LTFS11331E Failed to load the cartridge (ltfs_load_tape).
18d6e LTFS31258I Cannot get remaining capacity: failed to parse the log page
18d6e LTFS12030E Cannot get capacity data: backend call failed (-21704).
18d6e LTFS11999E Cannot load the medium: failed to get capacity data (-21704).
18d6e LTFS14075E Cannot set up tape drive.
18d6e LTFS31213I Error on modeselect: (pass18:mps0:0:83:0): MODE SELECT(10). CDB: 55 00 00 00 00 00 00 00 30 00
(pass18:mps0:0:83:0): CAM status: SCSI Status Error
(pass18:mps0:0:83:0): SCSI status: Check Condition
(pass18:mps0:0:83:0): SCSI sense: ILLEGAL REQUEST asc:24,0 (Invalid field in CDB)
(pass18:mps0:0:83:0): Command byte 1 bit 4 is invalid
 (-20501) HUJ527449P
18d6e LTFS17155E Cannot set WRITE MODE: Mode Select for Device Configuration Extension failed (-20501).
18d6e LTFS31213I Error on modesense: (pass18:mps0:0:83:0): MODE SENSE(10). CDB: 5a 00 24 00 00 00 00 00 28 00
(pass18:mps0:0:83:0): CAM status: SCSI Status Error
(pass18:mps0:0:83:0): SCSI status: Check Condition
(pass18:mps0:0:83:0): SCSI sense: ILLEGAL REQUEST asc:24,0 (Invalid field in CDB)
(pass18:mps0:0:83:0): Command byte 2 bit 5 is invalid
 (-20501) HUJ527449P
root@pzfjails:~/ltfs # ltfs -o device_list
LTFS9015W Setting the locale to 'en_US.UTF-8'. If this is wrong, please set the LANG environment variable before starting ltfs.
18e8d LTFS14000I LTFS starting, LTFS version 2.4.3.2 (Prelim), log level 2.
18e8d LTFS14058I LTFS Format Specification version 2.4.0.
18e8d LTFS14104I Launched by "ltfs -o device_list".
18e8d LTFS14105I This binary is built for FreeBSD.
18e8d LTFS14106I GCC version is FreeBSD Clang 10.0.1 (git@github.com:llvm/llvm-project.git llvmorg-10.0.1-0-gef32c611aa2).
18e8d LTFS17087I Kernel version: FreeBSD 12.2-STABLE r369131 GENERIC
.
18e8d LTFS17085I Plugin: Loading "cam" tape backend.
Tape Device list:.

**Comment by ia2115** [comment]
and on lto-6 tape on tandberg (freebsd 12.2)

# ltfs -o devname=/dev/sa0 /mnt/lto6
LTFS9015W Setting the locale to 'en_US.UTF-8'. If this is wrong, please set the LANG environment variable before starting ltfs.
188a6 LTFS14000I LTFS starting, LTFS version 2.4.3.2 (Prelim), log level 2.
188a6 LTFS14058I LTFS Format Specification version 2.4.0.
188a6 LTFS14104I Launched by "ltfs -o devname=/dev/sa0 /mnt/lto6".
188a6 LTFS14105I This binary is built for FreeBSD.
188a6 LTFS14106I GCC version is FreeBSD Clang 10.0.1 (git@github.com:llvm/llvm-project.git llvmorg-10.0.1-0-gef32c611aa2).
188a6 LTFS17087I Kernel version: FreeBSD 12.2-STABLE r369131 GENERIC
.
188a6 LTFS14063I Sync type is "time", Sync time is 300 sec.
188a6 LTFS17085I Plugin: Loading "cam" tape backend.
188a6 LTFS17085I Plugin: Loading "unified" iosched backend.
188a6 LTFS14095I Set the tape device write-anywhere mode to avoid cartridge ejection.
188a6 LTFS31223I Opening a device through CAM driver (/dev/sa0)
188a6 LTFS31228I Product ID is 'LTO-6 HH'
188a6 LTFS31229I Vendor ID is TANDBERG
188a6 LTFS31232I Firmware revision is 3579
188a6 LTFS31233I Drive serial is HUJ527449P
188a6 LTFS17160I Maximum device block size is 131072.
188a6 LTFS11330I Loading cartridge.
188a6 LTFS31258I Cannot get remaining capacity: failed to parse the log page
188a6 LTFS12030E Cannot get capacity data: backend call failed (-21704).
188a6 LTFS11999E Cannot load the medium: failed to get capacity data (-21704).
188a6 LTFS11331E Failed to load the cartridge (ltfs_load_tape).
188a6 LTFS31258I Cannot get remaining capacity: failed to parse the log page
188a6 LTFS12030E Cannot get capacity data: backend call failed (-21704).
188a6 LTFS11999E Cannot load the medium: failed to get capacity data (-21704).
188a6 LTFS14075E Cannot set up tape drive.
188a6 LTFS31213I Error on modeselect: (pass18:mps0:0:83:0): MODE SELECT(10). CDB: 55 00 00 00 00 00 00 00 30 00
(pass18:mps0:0:83:0): CAM status: SCSI Status Error
(pass18:mps0:0:83:0): SCSI status: Check Condition
(pass18:mps0:0:83:0): SCSI sense: ILLEGAL REQUEST asc:24,0 (Invalid field in CDB)
(pass18:mps0:0:83:0): Command byte 1 bit 4 is invalid
 (-20501) HUJ527449P
188a6 LTFS17155E Cannot set WRITE MODE: Mode Select for Device Configuration Extension failed (-20501).
188a6 LTFS31213I Error on modesense: (pass18:mps0:0:83:0): MODE SENSE(10). CDB: 5a 00 24 00 00 00 00 00 28 00
(pass18:mps0:0:83:0): CAM status: SCSI Status Error
(pass18:mps0:0:83:0): SCSI status: Check Condition
(pass18:mps0:0:83:0): SCSI sense: ILLEGAL REQUEST asc:24,0 (Invalid field in CDB)
(pass18:mps0:0:83:0): Command byte 2 bit 5 is invalid
 (-20501) HUJ527449P

**Comment by ia2115** [comment]
this can be closed linux Debian 10 is running now with tapedrives.

**Comment by xosevp** [comment]
This bug can be closed, since  support of Tandberg drives  was included in ab7d0f09ba5f133966c077e81eab23eb5876238a / bug #299
